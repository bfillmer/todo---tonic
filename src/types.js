
// ROUTES
export const ROUTE_TODOS = 'route/ROUTE_TODOS'

// TODOS
export const ADDED_TODO = 'todo/ADDED_TODO'
export const REMOVED_TODO = 'todo/REMOVED_TODO'
export const COMPLETED_TODO = 'todo/COMPLETED_TODO'

// STATUSES
export const COMPLETED = 'COMPLETED'
export const IN_PROGRESS = 'IN_PROGRESS'
