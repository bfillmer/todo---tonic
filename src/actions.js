
import cuid from 'cuid'
import {createAction} from 'redux-actions'

import * as types from 'types'

// TODOS
const todo = (
  text = '',
  id = cuid(),
  status = types.IN_PROGRESS
) => ({id, text, status})

export const addTodo = createAction(types.ADDED_TODO, todo)

export const removeTodo = createAction(types.REMOVED_TODO)

export const completeTodo = createAction(types.COMPLETED_TODO)
