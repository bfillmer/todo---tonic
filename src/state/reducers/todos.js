
import {handleActions} from 'redux-actions'

import {ADDED_TODO, REMOVED_TODO, COMPLETED_TODO, COMPLETED} from 'types'

const assign = Object.assign

export const initialState = {
  byId: {},
  allIds: []
}

const addTodo = (state, {payload}) => {
  const {id} = payload
  const byId = assign({}, state.byId, {
    [id]: payload
  })
  const allIds = [...state.allIds, id]
  return {
    byId,
    allIds
  }
}

const removeTodo = (state, {payload: id}) => {
  const byId = assign({}, state.byId)
  delete byId[id]
  const allIds = state.allIds.filter(i => i !== id)
  return {
    byId,
    allIds
  }
}

const completeTodo = (state, {payload: id}) => assign({}, state, {
  byId: assign({}, state.byId, {
    [id]: assign({}, state.byId[id], {status: COMPLETED})
  })
})

export const reducer = handleActions({
  [ADDED_TODO]: addTodo,
  [REMOVED_TODO]: removeTodo,
  [COMPLETED_TODO]: completeTodo
}, initialState)
