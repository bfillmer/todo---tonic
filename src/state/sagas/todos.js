
import {call, put, select, takeEvery} from 'redux-saga/effects'

import {ADDED_TODO, REMOVED_TODO, COMPLETED_TODO} from 'types'
import {getTodoState} from 'selectors'
import {hydrateTodos} from 'actions'

const ls = window.localStorage
const {stringify, parse} = JSON

const KEY = 'tandt-todos'

// PERSIST TODOS TO LOCALSTORAGE
// Actions that we want to persist our todo state on.
const persistTypes = [ADDED_TODO, REMOVED_TODO, COMPLETED_TODO]

function * persistToLocalStorage () {
  const todoState = yield select(getTodoState)
  yield ls.setItem(KEY, stringify(todoState))
}

export function * persist () {
  yield takeEvery(persistTypes, persistToLocalStorage)
}

// HYDRATE TODOS FROM LOCALSTORAGE
export function * hydrate () {
  const localStorageTodos = yield parse(ls.getItem(KEY))
  // @TODO Ideally confirm basic state shape.
  if (localStorageTodos !== null) {
    const action = yield call(hydrateTodos, localStorageTodos)
    yield put(action)
  }
}
