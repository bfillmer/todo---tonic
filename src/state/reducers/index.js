
import {combineReducers} from 'redux'

import {reducer as location} from 'state/routes'
import {reducer as todos} from 'state/reducers/todos'

export const reducers = combineReducers({
  location,
  todos
})
