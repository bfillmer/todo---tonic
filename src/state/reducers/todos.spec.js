/* global describe, expect, test */

import {ADDED_TODO, REMOVED_TODO, COMPLETED_TODO} from 'types'
import {addTodo, removeTodo, completeTodo} from 'actions'

import {initialState, reducer} from 'state/reducers/todos'

describe('[UNIT] Todo Reducer', () => {
  const firstTodo = addTodo('A simple todo item.', 1)
  const secondTodo = addTodo('Another todo item.', 2)
  const stateWithSingleTodo = reducer(initialState, firstTodo)
  const stateWithTwoTodos = reducer(stateWithSingleTodo, secondTodo)

  test(ADDED_TODO, () => {
    expect(stateWithSingleTodo).toMatchSnapshot()
    expect(stateWithTwoTodos).toMatchSnapshot()
  })
  test(REMOVED_TODO, () => {
    const removeAction = removeTodo(2)
    const stateAfterRemove = reducer(stateWithTwoTodos, removeAction)
    expect(stateAfterRemove).toEqual(stateWithSingleTodo)
  })
  test(COMPLETED_TODO, () => {
    const completeAction = completeTodo(1)
    const stateAfterComplete = reducer(stateWithTwoTodos, completeAction)
    expect(stateAfterComplete).toMatchSnapshot()
  })
})
