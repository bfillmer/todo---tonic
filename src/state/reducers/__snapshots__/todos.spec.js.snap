// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[UNIT] Todo Reducer todo/ADDED_TODO 1`] = `
Object {
  "allIds": Array [
    1,
  ],
  "byId": Object {
    "1": Object {
      "id": 1,
      "status": "IN_PROGRESS",
      "text": "A simple todo item.",
    },
  },
}
`;

exports[`[UNIT] Todo Reducer todo/ADDED_TODO 2`] = `
Object {
  "allIds": Array [
    1,
    2,
  ],
  "byId": Object {
    "1": Object {
      "id": 1,
      "status": "IN_PROGRESS",
      "text": "A simple todo item.",
    },
    "2": Object {
      "id": 2,
      "status": "IN_PROGRESS",
      "text": "Another todo item.",
    },
  },
}
`;

exports[`[UNIT] Todo Reducer todo/COMPLETED_TODO 1`] = `
Object {
  "allIds": Array [
    1,
    2,
  ],
  "byId": Object {
    "1": Object {
      "id": 1,
      "status": "COMPLETED",
      "text": "A simple todo item.",
    },
    "2": Object {
      "id": 2,
      "status": "IN_PROGRESS",
      "text": "Another todo item.",
    },
  },
}
`;
